src/passes/J2CLOpts.cpp

```diff
@@ -30,9 +30,48 @@ namespace wasm {
 
 namespace {
 
+using AssignmentCountMap = std::unordered_map<Name, Index>;
+using TrivialFunctionMap = std::unordered_map<Name, Expression*>;
+
 bool isOnceFunction(Function* f) { return f->name.hasSubstring("_<once>_"); }
 
-using AssignmentCountMap = std::unordered_map<Name, Index>;
+// Returns the function body if it is a trivial function, null otherwise.
+Expression* getTrivialFunctionBody(Function* func) {
+  auto* body = func->body;
+
+  // Only consider trivial the following instructions which can be safely
+  // inlined and note that their size is at most 2.
+  if (body->is<Nop>() || body->is<GlobalGet>() || body->is<Const>() ||
+      (body->is<Call>() && body->dynCast<Call>()->operands.size() == 0) ||
+      (body->is<GlobalSet>() &&
+       body->dynCast<GlobalSet>()->value->is<Const>())) {
+    return body;
+  }
+  return nullptr;
+}
+
+// Adds the function to the map if it is trivial.
+void maybeCollectTrivialFunction(Module* module,
+                                 Function* func,
+                                 TrivialFunctionMap& trivialFunctionMap) {
+  auto* body = getTrivialFunctionBody(func);
+  if (body == nullptr) {
+    return;
+  }
+
+  trivialFunctionMap[func->name] = body;
+}
+
+// Cleans up a once function that has been modified in the hopes it
+// becomes trivial.
+void cleanupFunction(Module* module, Function* func) {
+  PassRunner runner(module);
+  runner.add("precompute-propagate");
```

Comment:
how does precompute and precompute-propagate differ?

Reply:
I left `precompute-propagate` by mistake. In our scenarios `precompute` is enought. `precompute-propagate`  propagates the compile time constants through locals. I haven't seen this scenario in our once functions, but it is possible in `$clinit`  since that is code written by the user.